Rename billingData to billingAddress

nielslange opened the pull request:
Fixes #5811

In our codebase, we were using `billingData` and `shippingAddress`, which both contains the address object. To increase the maintainability of the codebase, we changed `billingData` to `billingAddress`. 

As some extensions are already using `billingData` via `useStoreCart`, this PR ensures that the change is backwards compatible.

## Testing

### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [x] Unit tests
  * [ ] E2E tests

For automatic testing, you can either run the entire unit tests using `npm t`, the individual unit test that covers this PR using `npm t -- assets/js/base/context/hooks/cart/test/use-store-cart.js` and the individual integration test using `npm t -- npm t -- slots.js`.

### User Facing Testing

#### Regular payments

1. Place a test order with the default payment method, e.g. Cash on Delivery.
2. Place a test order with an external payment gateways, e.g. [WooCommerce Stripe Payment Gateway](https://wordpress.org/plugins/woocommerce-gateway-stripe/).

#### Deprecation gate

1. Install the plugin [woocommerce-test-extension.zip](https://github.com/woocommerce/woocommerce-blocks/files/8880452/woocommerce-test-extension.zip)
2. Go to the Checkout page and open the console. (https://kb.mailster.co/how-can-i-open-the-browsers-console/ contains instructions on how to open the console.)
3. Verify that the warning `billingData is deprecated. Please use billingAddress instead. See: https://github.com/woocommerce/woocommerce-blocks/pull/6369` is visible.

<img width="839" alt="Screenshot 2022-06-10 at 18 18 40" src="https://user-images.githubusercontent.com/3323310/173108806-d2dfafd7-bdac-4a7e-95ea-29337260738f.png">

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

nielslange committed:
Rename billingData to billingAddress

nielslange committed:
Add unit test to ensure billingData remains accessible

senadir committed:
add integration tests for slots

wavvves commented on assets/js/base/context/providers/cart-checkout/payment-methods/use-payment-method-registration.ts:
```diff
@@ -63,7 +63,7 @@ const usePaymentMethodRegistration = (
 		cart,
 		cartTotals,
 		cartNeedsShipping,
-		billingData,
```
After a bit of reviewing with @opr, we found this is being used by registerPaymentMethodExtensionCallbacks (PR and more info [here](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/4668)) so billingData should probably still being carried up to this point.

At this point, I wonder if we should weigh all the work involved against the task purpose of just renaming this for consistency 🤔

nielslange replied:
> After a bit of reviewing with @opr, we found this is being used by registerPaymentMethodExtensionCallbacks (PR and more info https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/4668) so billingData should probably still being carried up to this point.

Thanks for pointing out this section. I created a quick prototype and can see this problem.

> At this point, I wonder if we should weigh all the work involved against the task purpose of just renaming this for consistency 🤔

Personally, I think it's worth it as it increases the readability on the long term. For example, we use `shippingData` and `shippingAdress`. What we called `billingData` is not similar to `shippingData` but to `shippingAdress`.

nielslange replied:
Thanks for pointing this out. I've added `billingData` back to `usePaymentMethodRegistration` for backwards compatibility.

wavvves commented on assets/js/base/context/providers/cart-checkout/payment-methods/use-payment-method-registration.ts:
```diff
@@ -63,7 +63,7 @@ const usePaymentMethodRegistration = (
 		cart,
 		cartTotals,
 		cartNeedsShipping,
-		billingData,
```
After a bit of reviewing with @opr, we found this is being used by registerPaymentMethodExtensionCallbacks (PR and more info [here](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/4668)) so billingData should probably still being carried up to this point.

At this point, I wonder if we should weigh all the work involved against the task purpose of just renaming this for consistency 🤔

opr replied:
@wavvves I see you approved the PR but the highlighted code you mentioned is going to cause errors with payment methods if they try to access `billingData` won't it?

wavvves replied:
@opr billingData support was added by @nielslange on the latest commit, can you have a look and see if you spot any problem?

wavvves commented on assets/js/base/context/hooks/cart/test/use-store-cart.js:
```diff
@@ -39,6 +39,19 @@ describe( 'useStoreCart', () => {
 		cartIsLoading: false,
 		cartItemErrors: [],
 		cartErrors: [],
+		billingData: {
```
@opr suggested we could also use [this](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get) instead of accessing the property directly. That way we could warn about deprecated usage, something along the lines of this maybe

![image](https://user-images.githubusercontent.com/17236129/167125401-5ac06c84-c7bf-4d45-8453-a0ef7cee0da8.png)

nielslange replied:
@wavvves I see that you referred to the file `assets/js/base/context/hooks/cart/test/use-store-cart.js`, which only contains the test. I assume that the deprecation gate should be added to `assets/js/base/context/providers/cart-checkout/payment-methods/use-payment-method-registration.ts` instead, correct?

wavvves replied:
Yes, sorry my bad, I picked up the wrong file after a search. Yes, I didn't want to mention the test file, thanks for noticing 👍

nielslange committed:
Keep billingData in usePaymentMethodRegistration for backwards compatibility

nielslange committed:
Gate `billingData` in deprecation gate

nielslange committed:
Merge branch 'trunk' into update/5811-rename-billingData-to-billingAddress-internally

nielslange commented:
@wavvves I addressed your feedback from https://github.com/woocommerce/woocommerce-blocks/pull/6369#pullrequestreview-964534360. Would you mind reviewing this PR again?

opr commented on assets/js/base/context/providers/cart-checkout/payment-methods/use-payment-method-registration.ts:
```diff
@@ -75,7 +76,14 @@ const usePaymentMethodRegistration = (
-			billingData,
+			get billingData() {
+				/* eslint no-console: ["error", { allow: ["warn"] }] */
+				console.warn(
```
Could we use https://developer.wordpress.org/block-editor/reference-guides/packages/packages-deprecated/ here?

nielslange replied:
@opr I changed the section before to:

```ts
get billingData() {
  // prettier-ignore
  deprecated(
    'billingData',
    {
      alternative: 'billingAddress',
      plugin: 'woocommerce-gutenberg-products-block',
      link:
        'https://github.com/woocommerce/woocommerce-blocks/pull/6369',
    }
  );
  return this.billingAddress;
},
```

Is that what you had in mind?

nielslange committed:
Replace deprecation call

nielslange committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-blocks into update/5811-rename-billingData-to-billingAddress-internally

# Conflicts:
#	package-lock.json

opr reviewed (approved):
thanks Niels, I tested and it works! 🙌🏼 

The user-facing testing instructions in this PR are incorrect, a user won't be checking out PRs or adding any code to their site, could we come up with an alternative testing strategy, for example using and verifying there are no errors (this plugin accesses `billingData` on this line: https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/extend-cart-checkout-block/src/js/filters.js.mustache#L17) 

[woocommerce-test-extension.zip](https://github.com/woocommerce/woocommerce-blocks/files/8880388/woocommerce-test-extension.zip)
 
I approved so you can merge, but please update the instructions!
